Allow options dialog to use a tree structure for tab list

nyalldawson opened the pull request:
This PR sets up a framework for options dialogs to use a structured tree to list the available pages, instead of just a flat list (the current behaviour). The intention here is to help improve the usability of the rapidly growing global options dialog, by setting up a tree structure which groups similar themed pages together to aid discoverability. Right now it only groups the the CRS and Transformations items together, but I have upcoming PRs which will introduce additional pages into this dialog so I'm looking for comments on this change first. (and at some stage I'd like to move "Custom Projections" into this global options dialog too, so that's another item which would sit nicely in this same group!)

Here's how it looks right now:

![image](https://user-images.githubusercontent.com/1829991/126744008-a04fa519-dae2-4421-97fd-ee768a5a8cf7.png)

(@nirvn I'll need some assistance with the stylesheet theming here -- I can't work out why that blue bar is showing in the margin!)

TODO:

- [x] fix item hiding when a search term is entered in the dialog
- [x] fix stylesheet
- [x] allow pages created via factories to have tree structures

nyalldawson committed:
Allow options dialog to use a tree structure for tab list instead
of a flat list widget

nirvn commented on src/app/qgisappstylesheet.cpp:
```diff
@@ -152,6 +152,23 @@ void QgisAppStyleSheet::buildStyleSheet( const QMap<QString, QVariant> &opts )
+                             "    color: white;"
+                             "    padding: %1px;"
+                             "}"
+                             "QTreeView#mOptionsTreeView::item::selected {"
```
```suggestion
                             "QTreeView#mOptionsTreeView::item::selected, QTreeView#mOptionsTreeView::branch::selected {"
```

![image](https://user-images.githubusercontent.com/1728657/126926541-b2b85565-4627-4c9e-adb6-c479686a5b94.png)

nirvn replied:
(@nyalldawson , ping)

nyalldawson replied:
thank you!

nirvn commented:
@nyalldawson , I really like where this is going. ATM, the only thing I feel isn't optimal is the lack of any feedback when single left clicking on a parent node that contains children. It should either auto-expand the parent node on single left click, or auto-expand and focus on the first child.

nyalldawson committed:
Fix stylesheet

nyalldawson committed:
Working search for items when using options tree views

nyalldawson committed:
Expand all children on open

nyalldawson committed:
Show search hits for tables and tree/table headers

nyalldawson committed:
Don't search html tags when searching text in options label widgets

nyalldawson committed:
Update and restore debug guard code

nyalldawson committed:
Resurrect screenshot code

nyalldawson committed:
Factories can specify a path to add their pages to the options tree

nyalldawson committed:
Expand all when clearing search

nyalldawson committed:
Fix and simplify options locator widget assertions

nyalldawson committed:
When searching for options pages, we should also show those
where the page name matches the search term

Otherwise searching for eg "python" won't show the python console
settings page.

nyalldawson commented:
This is ready for merge now!

nyalldawson commented:
@nirvn 
> ATM, the only thing I feel isn't optimal is the lack of any feedback when single left clicking on a parent node that contains children. It should either auto-expand the parent node on single left click, or auto-expand and focus on the first child.

I'm going to defer that for now. I feel like both these options are more of a short-term thing, when ultimately we'd go with Pycharm's approach of auto-creating a parent page:

![image](https://user-images.githubusercontent.com/1829991/127433903-ad983b51-5c92-49f3-95a2-b657e353b574.png)

nirvn reviewed (approved):
I'm sorry, I just can't find a fault in there, I guess you'll just have to merge :wink:
